### PR TITLE
winch: Add support for parametric instructions

### PR DESCRIPTION
**Pull request by saulecabrera**
This commit introduces support for the `drop` and `select` instructions.

Additionally, it refactors the `CodeGenContext::drop_last` implementation, enhancing flexibility for callers to determine the handling of elements to be dropped. This refactoring simplifies scenarios where a `Memory` entry is at the top of the stack.

### CODE REVIEW

File: winch/codegen/src/codegen/call.rs [hunk 1]
```diff
@@ -165,7 +165,28 @@ impl<'a> FnCall<'a> {
...
     fn post_call<M: MacroAssembler>(&self, masm: &mut M, context: &mut CodeGenContext, size: u32) {
         masm.free_stack(self.call_stack_space + size);
-        context.drop_last(self.abi_sig.params.len());
+        // Only account for registers given that any memory entries
+        // consumed by the call (assigned to a register or to a stack
+        // slot) were freed by the previous call to
+        // `masm.free_stack`, so we only care about dropping them
+        // here.
+        //
+        // NOTE / TODO there's probably a path to getting rid of
```
Comment (saulecabrera) [inline winch/codegen/src/codegen/call.rs:174]: I'm planning on doing this refactor, but I didn't want to introduce it as part of this change.

File: winch/codegen/src/visitor.rs [hunk 2]
```diff
@@ -637,6 +639,40 @@ where
...
+            _ => {}
+        });
+    }
+
+    fn visit_select(&mut self) {
+        let cond = self.context.pop_to_reg(self.masm, None);
+        let val2 = self.context.pop_to_reg(self.masm, None);
+        let val1 = self.context.pop_to_reg(self.masm, None);
+        let done = self.masm.get_label();
+        self.masm.branch(
```
Comment (cfallin) [inline winch/codegen/src/visitor.rs:655]: I'm somewhat surprised that select is being implemented with actual local control flow here; is there any reason we can't use a conditional move? That's the usual lowering in e.g. Cranelift (and avoiding actual branches is sometimes important at the guest-code level for Spectre mitigations).

File: winch/codegen/src/codegen/call.rs (hunk 1 above)
Comment (cfallin) [reply winch/codegen/src/codegen/call.rs:174]: It's not entirely clear what's going on in the diff here (i.e. why things changed): is it that the registers weren't actually freed before?

**Review by cfallin — COMMENTED** [review]
Generally looks good, thanks! Just a few questions below.

File: winch/codegen/src/visitor.rs (hunk 2 above)
Comment (saulecabrera) [reply winch/codegen/src/visitor.rs]: > is there any reason we can't use a conditional move? 

Good catch; no particular reason why we can't use a conditional move here, it's just that I didn't consider it initially, but there are very good arguments to do so. I've refactored this code to use conditional moves instead in https://github.com/bytecodealliance/wasmtime/pull/6912/commits/c56125cd833ff8107b285f4a2872d0930c014eef

File: winch/codegen/src/codegen/call.rs (hunk 1 above)
Comment (saulecabrera) [reply winch/codegen/src/codegen/call.rs:174]: The registers were actually freed, implicitly as part of the previous implementation of the `drop_last` function, with the new implementation, the processing of certain values like registers and memory are explicit and up to the caller. I noticed that this change to the `drop_last` function has the potential to reduce a bit the boilerplate in calls, hence the `NOTE / TODO` comment. Hope this clarifies things up.

**Review by cfallin — APPROVED** [review]
LGTM!

**Pull request merged**